audits.toml

```diff
@@ -885,6 +885,60 @@ criteria = "safe-to-deploy"
 version = "1.0.0"
 notes = "No unsafe usage or ambient capabilities. No functionality in it beyond a #[test]. "
 
+[[audits.android-activity]]
+who = "Robert Bragg <opensource@embark-studios.com>"
+criteria = "safe-to-deploy"
+version = "0.4.1"
+notes = """Some unsafe usage for JNI/FFI, such as implementing extern \"C\" functions for
+NativeActivity and to use the `ndk_sys` FFI bindings for the Android NDK libraries.
+
+The GameActivity backend depends on around 2k lines of third-party C/C++ code from Google
+as well as around 500 lines of C++ code for the GameText (input method) support.
+The C/C++ code is compiled with the `cc` crate.
+
+Although I have reviewed all of the C/C++ code for GameActivity + GameText there
+could be unknown soundness issues in there or potentially in any of the Android
+NDK APIs used, which are generally also implemented in C/C++.
+
+Written by Robert Bragg who now works at Embark Studios.
+"""
+
+[[audits.jni]]
+who = "Robert Bragg <opensource@embark-studios.com>"
+criteria = "safe-to-deploy"
+version = "0.21.1"
+notes = """Aims to provide a safe JNI (Java Native Interface) API over the
+unsafe `jni_sys` crate.
+
+This is a very general FFI abstraction for Java VMs with a lot of unsafe code
+throughout the API. There are almost certainly some edge cases with its design
+that could lead to unsound behaviour but it should still be considerably safer
+than working with JNI directly.
+
+A lot of the unsafe usage relates to quite-simple use of `from_raw` APIs to
+construct or cast wrapper types (around JNI pointers) which are fairly
+straight-forward to verify/trust in context.
+
+Some unsafe code has good `// # Safety` documentation (this has been enforced for
+newer code) but a lot of unsafe code doesn't document invariants that are
+being relied on.
+
+The design depends on non-trivial named lifetimes across many APIs to associate
+Java local references with JNI stack frames.
+
+The crate is not very actively maintained and was practically unmaintained for
+over a year before the 0.20 release.
+
+Robert Bragg who now works at Embark Studios became the maintainer of this
+crate in October 2022.
+
+In the process of working on the `jni` crate since becoming maintainer it's
+worth noting that I came across multiple APIs that I found needed to be
+re-worked to address safety issues, including ensuring that APIs that are not
+implemented safely are correctly declared as `unsafe`.
```

Comment:
are there any open filed issues around this in the repo that we could link here as a quick suggestion for people to check out? not super common in audits, but have seen that be used specifically around potential unsound behavior or open other issues.

not the end of the world if don't have stuff though, this is quite well described. 

though wonder if we can mention something that it is actively used in production and that we haven't run into any issues in practice with it (to give a bit more confidence)?

Reply:
Here's one pretty good reference issue: https://github.com/jni-rs/jni-rs/issues/392

Another decent reference would be https://github.com/jni-rs/jni-rs/issues/396 and https://github.com/jni-rs/jni-rs/pull/400

or a smaller example of needing to mark something `unsafe`: https://github.com/jni-rs/jni-rs/issues/385

There are quite lot of projects depending on the crate in production, including some fairly high-profile projects: https://crates.io/crates/jni/reverse_dependencies

Reply:
nice! think you can just add the list of those issues to the audit text there of something like "related issues:"

Reply:
oh wait they were closed issues, I meant things like still open issues :)

Reply:
Ah, right. The things I was referring to in that snippet you quoted where things that did get addressed at the time, and those are some examples.

Can maybe find a good pointer for some open issues to be aware of a bit later yeah (on mobile atm)

Reply:
okey, I found a few open issues that seem worth highlighting and can append something like this to the notes:
```
# Some Notable Open Issues
- https://github.com/jni-rs/jni-rs/issues/422 - questions soundness of linking
  multiple versions of jni crate into an application, considering the use
  of (separately scoped) thread-local-storage to track thread attachments
- https://github.com/jni-rs/jni-rs/issues/405 - discusses the ease with which
  code may expose the JVM to invalid booleans with undefined behaviour
```

A pretty decent example of a project depending on the `jni` crate in production is the Signal messaging app: https://github.com/signalapp/libsignal/blob/main/rust/bridge/jni/Cargo.toml

Reply:
nice sounds good!

oh and Signal is definitely a big important user :smile: :heart:

Reply:
Yeah think that one counts :)